src/Database/LSMTree/Internal/IndexOrdinary.hs

```diff
@@ -0,0 +1,222 @@
+       )
+import Data.Primitive.ByteArray
+       (
+           ByteArray,
+           indexByteArray
+       )
+import Data.ByteString.Short
+       (
+           ShortByteString (ShortByteString)
+       )
+import Data.ByteString.Short qualified as ShortByteString
+       (
+           length
+       )
+import Data.Vector
+       (
+           Vector,
+           fromList,
+           length,
+           (!),
+           drop,
+           takeWhile,
+           findIndexR
+       )
+import Data.Vector.Primitive qualified as Primitive
+       (
+           Vector (Vector),
+           force,
+           null,
+           length,
+           take,
+           drop,
+           splitAt,
+       )
+import Database.LSMTree.Internal.Serialise
+       (
+           SerialisedKey (SerialisedKey')
+       )
+import Database.LSMTree.Internal.Entry
+       (
+           NumEntries (NumEntries)
+       )
+import Database.LSMTree.Internal.IndexCompact
+       (
+           PageNo (PageNo),
+           PageSpan (PageSpan)
+       )
+
+{-|
+    The version of the index and its serialisation format that is supported by
+    this module.
+-}
+supportedVersion :: Word32
+supportedVersion = 1
+
+{-|
+    A general-purpose fence pointer index.
+
+    An index is represented by a vector that maps the number of each page to the
+    key stored last in this page or, if the page is an overflow page, to the key
+    of the corresponding key–value pair. From this, the following constraints on
+    the vector follow:
+
+      * The vector must be non-empty.
+
+      * The elements of the vector must be non-decreasing.
+-}
+newtype IndexOrdinary = IndexOrdinary (Vector SerialisedKey)
+
+-- * Search
+
+{-|
+    Returns the span of pages that contain the key–value pair of a certain key
+    if there is such a pair, or an arbitrary, valid page span if there is not.
+-}
+search :: SerialisedKey -> IndexOrdinary -> PageSpan
+search key (IndexOrdinary lastKeys) = PageSpan (PageNo start) (PageNo end) where
+
+    start :: Int
+    start | protoStart < pageCount = protoStart
+          | otherwise              = maybe 0 succ $
+                                     findIndexR (/= lastKeys ! pred pageCount) $
+                                     lastKeys
+
+    end :: Int
+    end | protoStart < pageCount
+            = start + length (takeWhile (== key) (drop (succ start) lastKeys))
+        | otherwise
+            = pred pageCount
+
+    protoStart :: Int
+    protoStart = searchForProtoStart 0 (length lastKeys)
+
+    searchForProtoStart :: Int -> Int -> Int
+    searchForProtoStart first count
+        | count == 0               = first
+        | key <= lastKeys ! center = searchForProtoStart first frontCount
+        | otherwise                = searchForProtoStart (succ center) rearCount
+        where
+
+        frontCount, rearCount, center :: Int
+        frontCount = count `div` 2
+        rearCount  = pred (count - frontCount)
+        center     = first + frontCount
```

Comment:
I crafted the local search function `searchForProtoStart` specifically for finding the start of a page span and was under the impression that because of some specifics it isn’t appropriate as a general-purpose function. However, I realize now that it really just performs an ordinary binary search that finds the first occurrence of a value in case there are multiple occurrences. Therefore, it is sensible to turn it into sort of a library function. I’ve created #252, which calls for this.

Reply:
> The reason for me not using `vector-algorithms` is precisely that it only works with mutable vectors. I guess that it would indeed be possible to use it by employing `unsafeThaw`. However, I find it questionable to use an unsafe operation where there is a safe alternative, just because of limitations of current libraries.

In this case, it is a safe use of `unsafeThaw`: it is only thawed within a local scope where no alterations to the vector are made. Maybe it's a shortcoming of the library that they don't have a version on immutable vectors, but I think it's sensible to use here. IMO we care more about performance than strict safety in the `lsm-tree` library, and it's still safe to use `unsafeThaw`, because of reasoning stated above. Moreover, `vector-algorithms` is to some degree tested/benchmarked/battle-tested, so that's also an advantage

> I crafted the local search function searchForProtoStart specifically for finding the start of a page span and was under the impression that because of some specifics it isn’t appropriate as a general-purpose function. However, I realize now that it really just performs an ordinary binary search that finds the first occurrence of a value in case there are multiple occurrences. Therefore, it is sensible to turn it into sort of a library function. I’ve created https://github.com/IntersectMBO/lsm-tree/issues/252, which calls for this.

:+1: I think this may also be necessary to some degree even if we use `vector-algorithms`, depending on which element is returned by those search functions

Reply:
Let’s continue this discussion over at #256, where I’ve already created [a comment](https://github.com/IntersectMBO/lsm-tree/pull/256#issuecomment-2178870318) summarizing the state of this discussion. We should leave this thread open however, to remind ourselves that we actually have to use the search function of #256 in the code of this pull request.

Reply:
The corresponding change has been introduced in 378a844b737a6416ea5c789f646e939710007f04.